File: log4all/model/application.py
import logging

from bson import ObjectId


__author__ = 'Igor Maculan <n3wtron@gmail.com>'

_log = logging.getLogger(__name__)


class Application:
    def __init__(self, name, token=None, description=None, configuration=None):
        self._id = None
        self.name = name
        self.token = token
        self.description = description
        self.configuration = configuration
        if self.configuration is None:
            self.configuration = {
                'DEBUG': {
                    'archive': 0,
                    'delete': 30
                },
                'INFO': {
                    'archive': 30,
                    'delete': 60
                },
                'WARNING': {
                    'archive': 60,
                    'delete': 180
                },
                'ERROR': {
                    'archive': 90,
                    'delete': 365
                }
            }

    def __json__(self, request=None):
        json = {
            'name': self.name,
            'token': self.token,
            'description': self.description,
            'configuration': self.configuration
        }
        if self._id is not None:
            json['_id'] = self._id
        return json

    @staticmethod
    def from_bson(bson):
        app = Application(bson['name'], bson.get('token'), bson.get('description'), bson.get('configuration'))
        app._id = str(bson['_id'])
        return app

    @staticmethod
    def init(db):
        db.applications.ensure_index('_id', unique=True)
        db.applications.ensure_index('name', unique=True)

    def save(self, db):
        json = self.__json__()
        if self._id is not None:
            del json['_id']
            db.applications.update({'_id': ObjectId(self._id)}, json, w=1)
        else:
            id = db.applications.insert(json, w=1)
            if id is not None:
                self._id = str(id)


    @staticmethod
    def get(db, src_query={}):
        return Application.from_bson(db.applications.find_one(src_query))

    @staticmethod
    def delete(db, id):
        db.applications.remove(id)

    @staticmethod
    def search(db, src_query={}):
        for dbApp in db.applications.find(src_query):
            yield Application.from_bson(dbApp)

File: log4all/app/models/application.go
package models

import (
	"gopkg.in/mgo.v2"
	"gopkg.in/mgo.v2/bson"
)

type LevelConfiguration struct {
	ArchiveDays int `json:"archive"`
	DeleteDays  int `json:"delete"`
}

type ApplicationPermission struct {
	Username   string `json:"username" bson:"username"`
	Permission string `json:"permission" bson:"permission`
}

type RetentionConfiguration struct {
	Debug   LevelConfiguration `json:"debug" bson:"debug"`
	Info    LevelConfiguration `json:"info" bson:"info"`
	Warning LevelConfiguration `json:"warning" bson:"warning"`
	Error   LevelConfiguration `json:"error" bson:"error"`
}

type ApplicationConfiguration struct {
	Retention RetentionConfiguration `json:"retention" bson:"retention"`
	WriteSafe bool                   `json:"writeSafe" bson:"writeSafe"`
}

type Application struct {
	BasicModel    `bson:"-"`
	Id            bson.ObjectId            `json:"_id" bson:"_id"`
	Name          string                   `json:"name" bson:"name"`
	Description   string                   `json:"description" bson:"description"`
	Token         string                   `json:"token" bson:"token"`
	Configuration ApplicationConfiguration `json:"configuration" bson:"configuration"`
	Permissions   []ApplicationPermission  `json:"permissions" bson:"permissions"`
}

func CreateApplicationIndexes(db *mgo.Database) error {
	index := mgo.Index{
		Key:        []string{"name"},
		Unique:     true,
		DropDups:   true,
		Background: true,
	}
	return db.C("applications").EnsureIndex(index)
}

func (this *Application) Save(db *mgo.Database) error {
	this.Id = bson.NewObjectId()
	return db.C("applications").Insert(this)
}

func (this *Application) Update(db *mgo.Database, id string) error {
	app, err := GetApplicationById(db, id)
	if err == nil {
		err = db.C("applications").Update(bson.M{"_id": bson.ObjectIdHex(id)}, this)
		if this.Name != app.Name {
			//rename all logs for this application
			_, err = db.C("logs").UpdateAll(bson.M{"application": app.Name}, bson.M{"$set": bson.M{"application": this.Name}})
		}
	}
	return err
}

func DeleteApplication(db *mgo.Database, id string) error {
	app, err := GetApplicationById(db, id)
	if err == nil {
		err = db.C("applications").Remove(bson.M{"_id": bson.ObjectIdHex(id)})
	}
	if err == nil {
		_, err = db.C("logs").RemoveAll(bson.M{"application": app.Name})
	}
	return err
}

func GetApplicationById(db *mgo.Database, id string) (Application, error) {
	var result Application
	err := db.C("applications").Find(bson.M{"_id": bson.ObjectIdHex(id)}).One(&result)
	return result, err
}

func GetApplications(db *mgo.Database) ([]Application, error) {
	var result []Application
	err := db.C("applications").Find(bson.M{}).All(&result)
	return result, err
}

func GetApplicationByName(db *mgo.Database, name string) (*Application, error) {
	result := new(Application)
	err := db.C("applications").Find(bson.M{"name": name}).One(&result)
	return result, err
}

func GetApplicationByToken(db *mgo.Database, name string, token string) (*Application, error) {
	result := new(Application)
	err := db.C("applications").Find(bson.M{"name": name, "token": token}).One(&result)
	return result, err
}

func FindApplications(db *mgo.Database, query map[string]interface{}) ([]Application, error) {
	var result []Application
	err := db.C("applications").Find(query).All(&result)
	return result, err
}
